### PR TITLE
Tasks 

### LINKED ISSUE #37: Tasks
## Description

Adding tasks to the lib 

## The problem

Having to implement a task loop by yourself

## The solution

Implement it

## Checklist

- [x] - You have checked that there is no issue already mentioning this feature, open or closed
- [x] - Full details of the requested feature have been provided/shown

### PR DESCRIPTION
**Pull request by nav-github01001**
## Description
Closes #37 
 This PR adds tasks to simplify dev experience
PS: Excuse the wrong name of the commit. It is tasks instead of checks

## Checklist

- [x] - All code changes have been tested, if any
- [ ] - The documentation has been updated accordingly for any code changes
- [x] - This PR fixes an already existing issue
- [x] - This PR adds a new feature (new functions, methods, classes, parameters, etc.)
- [ ] - This PR fixes a bug
- [ ] - This PR is a breaking change (removes/renames functions, parameters etc.)
- [ ] - This PR is NOT a code change (changes Documentation files, README.md, CONTRIBUTING.md etc.)

### CODE REVIEW

File: EpikCord/__init__.py [hunk 1]
```diff
@@ -2279,6 +2280,23 @@ def latency(self):
...
+    def add_check(self, check:Callable, interval:Optional[int]=None, **kwargs):
+        async def full_check(check, interval=5, **kwargs):
+            task_start = False
+            if kwargs["start"] and kwargs["until"]:# Start when and do until
+                while True:
+                    if time() == kwargs["start"] or task_start == True:
+                        await check()
+                        task_start = True
+                        await asyncio.sleep(interval)
+
```
Comment (ZenoABC) [inline EpikCord/__init__.py:2292]: This looks inefficient way to do tasks, couldn't you think of something else ?

Comment (nav-github01001) [reply EpikCord/__init__.py:2318]: pls suggest another way if applicable

Comment (ZenoABC) [reply EpikCord/__init__.py:2318]: Is this tasks or checks ?

Comment (nav-github01001) [reply EpikCord/__init__.py:2318]: tasks... look at description i made a mistake

Comment (ZenoABC) [reply EpikCord/__init__.py:2318]: I don't know how to explain, but why not something like `@task(hour=1, minute=2)` ?, Then doing forever loop till cancel or the task errors, sleeping after each loop, and a way to tell max loops

Comment (nav-github01001) [reply EpikCord/__init__.py:2318]: i did 3 of what you told

Comment (nav-github01001) [reply EpikCord/__init__.py:2318]: i must commit that lol

Comment (ZenoABC) [reply EpikCord/__init__.py:2318]: What about remaining 3 ?

Comment (ZenoABC) [reply EpikCord/__init__.py:2318]: Anyways, i don't think using time() for starting will work, i don't know how to explain that part.

Comment (ZenoABC) [reply EpikCord/__init__.py:2318]: Or better yet, rethink about ways of doing this, take a look at dpy's way of doing it.

Comment (nav-github01001) [reply EpikCord/__init__.py:2318]: makes sense , will add the part

File: EpikCord/__init__.py [hunk 2]
```diff
@@ -2305,6 +2305,65 @@ def latency(self):
...
+            task_start = False
+            if kwargs.get("start") and kwargs.get("until"):  # Start when and do until
+                while True:
+                    current_time = time()
+
+                    if current_time >= float(kwargs["start"]) or task_start == True:
+                        try:
+                            await task()
+                        except Exception as e:
+                            raise TaskFailedError(f"Task failed due to {e}")
```
Comment (ZenoABC) [inline EpikCord/__init__.py:2339]: How about full traceback ?

File: EpikCord/__init__.py [hunk 3]
```diff
@@ -2305,6 +2305,65 @@ def latency(self):
...
+
+                        task_start = True
+                        await asyncio.sleep(interval)
+                    if current_time >= float(kwargs["until"]):
+                        break
+            elif kwargs.get("instances"):
+                instances = int(kwargs.get("instances"))
+                finished_instances = 0
```
Comment (ZenoABC) [inline EpikCord/__init__.py:2347]: Why is this hardcoded ?

Comment (nav-github01001) [reply EpikCord/__init__.py:2347]: wdym

File: EpikCord/__init__.py (hunk 2 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2339]: can you elaborate on that issue?

Comment (ZenoABC) [reply EpikCord/__init__.py:2339]: Instead of getting single line that won't be useful much, why not put full traceback there ?

File: EpikCord/__init__.py (hunk 3 above)
Comment (ZenoABC) [reply EpikCord/__init__.py:2347]: This is always set to 0, and since ur doing while instances < finished_instances, the if statement below will never run

File: EpikCord/__init__.py (hunk 2 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2339]: e is the full traceback

File: EpikCord/__init__.py (hunk 3 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2347]: sign messup

File: EpikCord/__init__.py (hunk 2 above)
Comment (ZenoABC) [reply EpikCord/__init__.py:2339]: It isn't.

File: EpikCord/__init__.py [hunk 4]
```diff
@@ -2305,6 +2305,65 @@ def latency(self):
     def average_latency(self):
         return sum(self.latencies) / len(self.latencies)
 
+    def add_task(self, task: Callable, interval: Optional[int] = None, **kwargs):
+        """Adds a background task (Tasks that run silently in the background)
+
+        This task can be used to monitor the bot's performance, change and send stuff routinely
+        through webhooks, etc
+
+        You can set the time of when it starts, stops and  add which arguments to give to the task
```
Comment (TheUntraceable) [inline EpikCord/__init__.py:2314]: Remove this whitespace

File: EpikCord/__init__.py [hunk 5]
```diff
@@ -2305,6 +2305,65 @@ def latency(self):
...
+
+        You may also define whether to run once or run indefinitely
+
+        Args:
+            task (Callable): The function to run on the background
+            interval (int, optional): The interval (in seconds) for a delay between when the instance of the task stops and starts again .Defaults to 5.
```
Comment (TheUntraceable) [inline EpikCord/__init__.py:2320]: The typehint should be like `interval: Optional[int]`, not that

**Review by TheUntraceable — CHANGES_REQUESTED** [review]

File: EpikCord/__init__.py (hunk 5 above)
Comment (nav-github01001) [reply EpikCord/__init__.py]: done

File: EpikCord/__init__.py (hunk 4 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2313]: done

File: EpikCord/__init__.py (hunk 1 above)
Comment (TheUntraceable) [reply EpikCord/__init__.py]: Maybe doing a better way such as
```py
async def task(self, *, seconds, minutes, hours, count):
    async def inner(func):
        time = calculate_time(seconds, minutes, hours)
        callback = None
        if count:
            async def callback():
                for i in range(count):
                    await func()
                    await sleep(time)
        else:
            async def callback():
                await func()
                await sleep(time)
                await asyncio.call_later(time, callback)
        asyncio.get_event_loop().run_until_complete()
```

File: EpikCord/__init__.py [hunk 6]
```diff
@@ -2305,6 +2305,63 @@ def latency(self):
     def average_latency(self):
         return sum(self.latencies) / len(self.latencies)
 
+    def add_task(self, task: Callable, interval: Optional[int] = None, **kwargs):
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2308]: `Callable` should be precised as `Callable[[...], None]`

File: EpikCord/__init__.py [hunk 7]
```diff
@@ -2305,6 +2305,63 @@ def latency(self):
...
+                    await asyncio.sleep(interval)
+
+            else:
+                while True:
+                    try:
+                        await task()
+                    except Exception as e:
+                        raise TaskFailedError(f"Task failed due to: {e}")
+
+                    await asyncio.sleep(interval)
```
[the remark this answers is not in the document]
Comment (TheUntraceable) [reply EpikCord/__init__.py]: > I should be put into a function or method
Agreed. I really thought from the start you should be put into a function or method.

[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py]: yessir

File: EpikCord/__init__.py [hunk 8]
```diff
@@ -17,7 +17,7 @@
 from importlib import import_module
 from inspect import iscoroutine
 from logging import getLogger
-from time import perf_counter_ns
+from time import perf_counter_ns, time
 from .type_enums import *
```
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py:21]: thats what i thought at first... Will implement the changes

File: EpikCord/__init__.py (hunk 6 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2308]: did not know that, excuse me for my dumbness

File: EpikCord/__init__.py (hunk 2 above)
Comment (Sigmanificient) [reply EpikCord/__init__.py:2337]: need to use `raise e` to get the full tb

File: EpikCord/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Jerry-py) [reply EpikCord/__init__.py:21]: It should

File: EpikCord/__init__.py [hunk 9]
```diff
@@ -17,7 +17,7 @@
 from importlib import import_module
 from inspect import iscoroutine
 from logging import getLogger
-from time import perf_counter_ns
+import time
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:20]: ...why?

File: EpikCord/__init__.py [hunk 10]
```diff
@@ -2305,6 +2305,56 @@ def latency(self):
...
+        async def full_task(task, **kwargs):
+            async def task_func(interval):
+                try:
+                    await task()
+                except Exception as e:
+                    raise TaskFailedError(f"Task failed due to: {e}")
+
+                await asyncio.sleep(interval)
+
+            interval = int(kwargs.get("interval", 5))
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2335]: could have a better support for passing `h`, `m`, `s`, ...
See dpy implementation

File: EpikCord/__init__.py [hunk 11]
```diff
@@ -2305,6 +2305,56 @@ def latency(self):
...
+            task_start = False
+            nb_instances = kwargs.get("instances")
+            if kwargs.get("start") and kwargs.get("until"):  # Start when and do until
+                while True:
```
Comment (Sigmanificient) [inline EpikCord/__init__.py:2339]: `While True` should be avoided

File: EpikCord/__init__.py (hunk 10 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2335]: Will be implementing support for h,m,s tmrw, I am not well

File: EpikCord/__init__.py (hunk 9 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:20]: ?

File: EpikCord/__init__.py (hunk 11 above)
Comment (nav-github01001) [reply EpikCord/__init__.py:2339]: I do not understand

File: EpikCord/__init__.py (hunk 9 above)
Comment (Sigmanificient) [reply EpikCord/__init__.py:20]: Not sure why you changed `from time import perf_counter_ns` to `import time`

File: EpikCord/__init__.py (hunk 10 above)
Comment (Sigmanificient) [reply EpikCord/__init__.py:2335]: tmrw? more like `d` for days?
Could be nicer to use plain keywords tho...

File: EpikCord/__init__.py (hunk 11 above)
Comment (Sigmanificient) [reply EpikCord/__init__.py:2339]: > I do not understand

```py
while True:
    ...
    if current_time >= float(kwargs["until"]):
        break
```

can't convert that to
```py
until = float(kwargs["until"])
while current_time < until:
    ...
```
?

File: EpikCord/__init__.py (hunk 9 above)
Comment (TheUntraceable) [reply EpikCord/__init__.py:20]: Same.

File: EpikCord/__init__.py (hunk 10 above)
Comment (TheUntraceable) [reply EpikCord/__init__.py:2335]: tmrw = Tomorrow
He meant to say he will do it tomorrow.

File: EpikCord/__init__.py (hunk 11 above)
Comment (TheUntraceable) [reply EpikCord/__init__.py:2339]: Yeah it can

Comment (nav-github01001) [reply EpikCord/__init__.py:2339]: ok, will implement imporovements

File: EpikCord/__init__.py [hunk 12]
```diff
@@ -4420,6 +4413,13 @@ def __init__(self, client):
         self._MARKDOWN_STOCK_REGEX = (
             rf"(?P<markdown>[_\\~|\*`]|{self._MARKDOWN_ESCAPE_COMMON})"
         )
+    
+    @staticmethod()
+    def get_time_since_epoch(year,month,day,hour,minute,second) -> float:
```
Comment (TheUntraceable) [inline EpikCord/__init__.py:4418]: Make them optional kwargs

File: EpikCord/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/__init__.py]: he meant that time can be used for some thing ele

File: EpikCord/ext/__init__.py [hunk 13]
```diff
@@ -0,0 +1 @@
+from .tasks import *
```
Comment (Sigmanificient) [inline EpikCord/ext/__init__.py:1]: That was very much needed thanks for making that init lighter

File: EpikCord/ext/tasks.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+import asyncio
+from ..exceptions import TaskFailedError
+
+_EPOCH = datetime.datetime(1970, 1, 1)
+
+
+class TimeParser:
+    @staticmethod
+    def get_time_since_epoch(
+        year: int, month: int, day: int, hour: int, minute: int, seconds: float
```
Comment (ZenoABC) [inline EpikCord/ext/tasks.py:13]: Why dont you make hour, minute and seconds optional ?

File: EpikCord/ext/__init__.py (hunk 13 above)
Comment (ZenoABC) [inline EpikCord/ext/__init__.py:1]: If you are gonna use this, make `__all__` in tasks.py atleast

**Review by ZenoABC — APPROVED** [review]
Other than that, looks good to me.

File: EpikCord/ext/tasks.py (hunk 14 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: oh

File: EpikCord/ext/tasks.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    ) -> float:
+        date = datetime.datetime(year, month, day, hour, minute, seconds)
+
+        delta = date - _EPOCH
+        return delta.total_seconds()
+
+    @staticmethod
+    def total_seconds(
+        year: int = 0,
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:31]: i dont the the point in having `month` & `year` honestly.
Plus note that a month != 30 as it varies between [28, 31]. Year also will not take leap into account. 

So, if you want to make a system that would work every X of a month, i would rather have something like a daily check, or a crontab implementation (which could be really useful).

As for year, ngl, it pretty much a joke

File: EpikCord/ext/tasks.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+class Tasks:
+    def __init__(self, client) -> None:
+        self.client = client
+
+    def add_task(
+        self,
+        task: Callable[..., None],
+        interval: Optional[int] = None,
+        start: Optional[float] = None,
+        until: Optional[float] = None,
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:55]: Would be cleaner using datetime obj directly

File: EpikCord/ext/tasks.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        **kwargs,
+    ):
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:57]: Kwargs for max execution? 

Like
```py
limit = 5
```
So the task run 5 times exactly.

File: EpikCord/ext/tasks.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+            until (optional[float]): The time to end, you may use `TimeParser.get_time_since_epoch` method to give the seconds since python epoch to this arg
+
+
+        """
+        kwargs["interval"] = interval
+        kwargs["start"] = start
+        kwargs["until"] = until
+
+        async def full_task(client, task, **kwargs):
+            async def task_func(client, interval):
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:80]: Since the `client` obj is accessed from outer scope, what the point of passing it again to `task_func`?

File: EpikCord/ext/tasks.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+                await asyncio.sleep(interval)
+
+            interval = int(kwargs.get("interval", 5))
+            task_start = False
+            nb_instances = kwargs.get("instances")
+            if kwargs.get("start") and kwargs.get("until"):  # Start when and do until
+                current_time = _time()
+                until = float(kwargs.get("until"))
+                if current_time >= float(kwargs.get("start")) or task_start:
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:94]: Maybe could flip the condition to avoid nesting, would simple `return` here works tho?

File: EpikCord/ext/tasks.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                while instances > finished_instances:
+                    await task_func(client, interval)
+
+            else:
+                await task_func(client, interval)
+
+        task = asyncio.get_event_loop().create_task(
+            full_task(self.client, task, **kwargs)
+        )
+        self.client.task.append(task)
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:110]: wtf is `self.client.task` and not `self.client.tasks`?

File: setup.cfg [hunk 21]
```diff
@@ -24,13 +24,16 @@ include_package_data = True
 packages = 
     EpikCord
     EpikCord.managers
+    EpikCord.ext
 install_requires = 
     aiohttp
 
-python_requires == ">=3.8"
+python_requires = >3.7
```
Comment (Sigmanificient) [inline setup.cfg:31]: Not sure about this?

File: EpikCord/ext/tasks.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
+import datetime
+from typing import Callable, Optional
+from time import time as _time
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:3]: Rename import should be avoid for clarity, i dont see where is the name conflct?

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 15 above)
Comment (ZenoABC) [reply EpikCord/ext/tasks.py:31]: True.

Comment (nav-github01001) [reply EpikCord/ext/tasks.py:31]: yeah, i wrote a comment abt it and i deleted it, it is a headache

File: EpikCord/ext/tasks.py (hunk 22 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:3]: will implement the changes just now

File: EpikCord/ext/tasks.py (hunk 19 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:94]: wdym

File: EpikCord/ext/tasks.py (hunk 15 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:31]: i will remove month and year then?

Comment (ZenoABC) [reply EpikCord/ext/tasks.py:31]: Yes.

File: EpikCord/ext/tasks.py (hunk 17 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:57]: we have a kwarg for that. "instances" but if you want i can rename it to limit

File: EpikCord/ext/tasks.py (hunk 18 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:80]: will change that in a new commit

File: EpikCord/ext/tasks.py (hunk 17 above)
Comment (ZenoABC) [reply EpikCord/ext/tasks.py:57]: "instances" is too vague, so i'd recommend a change.

Comment (nav-github01001) [reply EpikCord/ext/tasks.py:53]: ok

File: EpikCord/ext/tasks.py [hunk 23]
```diff
@@ -51,8 +47,8 @@ def add_task(
         self,
         task: Callable[..., None],
         interval: Optional[int] = None,
-        start: Optional[float] = None,
-        until: Optional[float] = None,
+        start: Optional[datetime.timedelta] = None,
+        until: Optional[datetime.timedelta] = None,
```
Comment (Sigmanificient) [inline EpikCord/ext/tasks.py:51]: Passing timedeltas type....? Why

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 23 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:51]: if i misunderstood your former change request?

Comment (Sigmanificient) [reply EpikCord/ext/tasks.py:51]: yeah...

Comment (nav-github01001) [reply EpikCord/ext/tasks.py:51]: i did, what was it meant to be

File: EpikCord/ext/tasks.py [hunk 24]
```diff
@@ -0,0 +1,107 @@
+import datetime
+from typing import Callable, Optional
+from time import time
+import asyncio
+from ..exceptions import TaskFailedError
+
+_EPOCH = datetime.datetime(1970, 1, 1)
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:7]: Surely for this you would want .timestamp()?

File: EpikCord/ext/tasks.py [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+    ):
+        """Adds a background task (Tasks that run silently in the background)
+
+        This task can be used to monitor the bot's performance, change and send stuff routinely
+        through webhooks, etc
+        You can set the time of when it starts, stops and  add which arguments to give to the task
+        You may also define whether to run once or run indefinitely
+
+        Args:
+            task (Callable): The function to run on the background
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:62]: Use the syntax of `task: Callable\n    Description`

File: EpikCord/ext/tasks.py [hunk 26]
```diff
@@ -0,0 +1,107 @@
...
+        async def full_task(client, task, **kwargs):
+            async def task_func(interval):
+                try:
+                    await task(client)
+                except Exception as e:
+                    raise TaskFailedError(f"Task failed due to: {e}")
+
+                await asyncio.sleep(interval)
+
+            interval = int(kwargs.get("interval", 5))
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:84]: This doesn't make sense to me. Surely you can just use the interval passed in the top level function?

File: EpikCord/ext/tasks.py [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+            task_start = False
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:85]: It'd make more sense to be named `task_started`, but eh.

File: EpikCord/ext/tasks.py (hunk 19 above)
Comment (TheUntraceable) [reply EpikCord/ext/tasks.py:90]: Flip it. For example
```py
if a == 12:
    if b == 13:
        if c == 14:

# Can be turned into
if a != 12:
    return
if b != 13:
    return
if c != 14:
    return

```

Comment (nav-github01001) [reply EpikCord/ext/tasks.py:90]: no need

**Review by TheUntraceable — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 24 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:7]: let it be

File: EpikCord/ext/tasks.py (hunk 17 above)
Comment (Sigmanificient) [reply EpikCord/ext/tasks.py:53]: > we have a kwarg for that. "instances" but if you want i can rename it to limit

Well it's isn't specified by default, meaning any completion tools would not suggest it.

File: EpikCord/ext/tasks.py (hunk 19 above)
Comment (Sigmanificient) [reply EpikCord/ext/tasks.py:90]: Could raise error on invalid parameter like

```py
add_task(
   ...
    start=datetime.now(),
    end=datetime.datetime(2001, 11, 12)
    ...
)  # => Should raise InvalidParameterError, or, InvalidTaskError, ...
```

**Review by Sigmanificient — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 19 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:90]: ok

Comment (Sigmanificient) [reply EpikCord/ext/tasks.py]: well we removed it lol

Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: yeah lmfao

File: EpikCord/ext/tasks.py [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+        You may also define whether to run once or run indefinitely
+
+        Args:
+            task (Callable): 
+            The function to run on the background
+            interval (Optional[int]): 
+            The interval (in seconds) for a delay between when the instance of the task stops and starts again.Defaults to 5.
+            You may use `TimeParser.total_seconds` method
+            instances (Optional[int]): 
+            The number of instances to start after delay. Usually unlimited
```
[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:52]: Ah yes

[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:52]: > Misleading description. How are you going to say the number of instances to start, usually unlimited? People will think the task will run like
> 
> ```python
> while True:
>     print("Task")
> ```

It is like that but with delays, i do not understand what is "misleading" here

[the remark this answers is not in the document]
Comment (Sigmanificient) [reply EpikCord/ext/tasks.py:53]: What he means by "misleading" is that people could think a task is run x time for every intervals

[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:53]: > What he means by "misleading" is that people could think a task is run x time for every intervals

I quite did not understand this
\

[the remark this answers is not in the document]
Comment (Sigmanificient) [reply EpikCord/ext/tasks.py:53]: People could understand
```py
while True:
     for i in range(instances):
         task()

    sleep(...)
```

Instead of
```py
while instance:
    task()
    sleep(...)
    instance -= 1

```

[the remark this answers is not in the document]
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:53]: I got it . Will make a fix now

File: EpikCord/ext/tasks.py [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+_EPOCH = datetime.datetime(1970, 1, 1)
+__all__ = (
+    "TimeParser",
+    "Tasks",
+)
+
+
+
+
+@staticmethod
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:15]: Isn't this only required for classes where the function doesn't need `self`

File: EpikCord/ext/tasks.py [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+    minute += hour * 60
+    seconds += minute * 60
+    return seconds
+
+
+class Tasks:
+    def __init__(self, client) -> None:
+        self.client = client
+
+    def add_task(
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:32]: Where is `limit`? It's in the docstring, but not in the signature for this function

File: EpikCord/ext/tasks.py [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+  
+        :param task: The function to run on the background
+        :type task: Callable[..., None]
+        :param interval:The interval (in seconds) for a delay between when the instance of the task stops and starts again.Defaults to 5.
+                        You may use `TimeParser.total_seconds` method, defaults to None
+        :type interval: Optional[int]
+        :param limit: The number of instances to start. Usually unlimited
+        :type limit: Optional[int] 
+        """        
+        kwargs["interval"] = interval
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:54]: Why is this needed?

File: EpikCord/ext/tasks.py [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+                try:
+                    await task(client, *args,**kwargs)
+                except Exception as e:
+                    raise TaskFailedError(f"Task failed due to: {e}")
+
+                await asyncio.sleep(interval)
+
+            interval = interval if interval <= 0 else 5
+            task_start = False
+            nb_instances = kwargs.get("limit")
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:67]: What does `nb_instances` mean?

File: EpikCord/ext/tasks.py [hunk 33]
```diff
@@ -0,0 +1,82 @@
...
+
+            if nb_instances:
+                instances = int(nb_instances)
+                finished_instances = 0
+                while instances > finished_instances:
+                    await task_func(client, interval)
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:73]: `finished_instances` isn't ever incremented, meaning this goes on forever :facepalm:

**Review by TheUntraceable — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 30 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:32]: limit is a kwarg 4 your information

Comment (TheUntraceable) [reply EpikCord/ext/tasks.py:32]: Make it explicit. If not then it's going to be passed into task callbacks

Comment (nav-github01001) [reply EpikCord/ext/tasks.py:32]: i did

File: EpikCord/ext/tasks.py (hunk 31 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: done

File: EpikCord/ext/tasks.py (hunk 32 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: done

File: EpikCord/ext/tasks.py (hunk 33 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:73]: done

File: EpikCord/ext/tasks.py (hunk 32 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: I just realized the var was useless, lemme remove it

File: EpikCord/ext/tasks.py (hunk 33 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:73]: Added increment, P.F.A latest commit

File: EpikCord/ext/tasks.py (hunk 31 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: sorry my bad, coding mistake

File: EpikCord/ext/tasks.py (hunk 30 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:32]: I made it explicit P.F.A latest commit

File: EpikCord/ext/tasks.py (hunk 29 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py]: Oops, 4got to remove it after deleting the parent class

File: EpikCord/ext/tasks.py [hunk 34]
```diff
@@ -0,0 +1,83 @@
+import datetime
+from typing import Callable, Optional
+import asyncio
+from ..exceptions import TaskFailedError
+
+_EPOCH = datetime.datetime(1970, 1, 1)
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:6]: Is this variable ever used?

File: EpikCord/ext/tasks.py [hunk 35]
```diff
@@ -0,0 +1,83 @@
...
+
+
+class Tasks:
+    def __init__(self, client) -> None:
+        self.client = client
+
+    def add_task(
+        self,
+        task: Callable[..., None],
+        interval: Optional[int] = None,
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:35]: make everything optional a kwarg, and maybe add a decorator for ease.

File: EpikCord/ext/tasks.py [hunk 36]
```diff
@@ -0,0 +1,83 @@
...
+        async def full_task(client, task, *args,**kwargs):
+            async def task_func(interval):
+                try:
+                    await task(client, *args,**kwargs)
+                except Exception as e:
+                    raise TaskFailedError(f"Task failed due to: {e}")
+
+                await asyncio.sleep(interval)
+
+            interval = interval if interval <= 0 else 5
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:66]: This can be changed to `interval = interval or 5`

File: EpikCord/ext/tasks.py [hunk 37]
```diff
@@ -0,0 +1,83 @@
...
+            task_start = False
+
+            if limit:
+                instances = int(nb_instances)
```
Comment (TheUntraceable) [inline EpikCord/ext/tasks.py:70]: Reference to a dead variable

**Review by TheUntraceable — CHANGES_REQUESTED** [review]

File: EpikCord/ext/tasks.py (hunk 35 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:35]: But then you said to make all vars explicit so that it doesnt get passed to the actual function

File: EpikCord/ext/tasks.py (hunk 34 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:6]: It is useless, forgot to remove it

File: EpikCord/ext/tasks.py (hunk 36 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:66]: makes sense, but what would you do if user pass interval = -5?(can be from typo)

File: EpikCord/ext/tasks.py (hunk 37 above)
Comment (nav-github01001) [reply EpikCord/ext/tasks.py:70]: oops
